### ISSUE TITLE
Identity (single element) array support in the SDK

### ISSUE THREAD

**Issue by mohammadfawaz** [FuelLabs/fuels-rs#233]
Example:

Sway code:
```rust
contract;

abi MyContract {
    fn small_array_identity(a: [u64; 1]) -> [u64; 1];
}

impl MyContract for Contract {
    fn small_array_identity(a: [u64; 1]) -> [u64; 1] {
        a
    }
}
```

Test:
```rust
use fuel_tx::{ContractId, Salt};
use fuels_abigen_macro::abigen;
use fuels_rs::prelude::*;
use fuels_signers::util::test_helpers;

// Load abi from json
abigen!(MyContract, "out/debug/proj-abi.json");

async fn get_contract_instance() -> (MyContract, ContractId) {
    // Deploy the compiled contract
    let salt = Salt::from([0u8; 32]);
    let compiled = Contract::load_sway_contract("./out/debug/proj.bin", salt).unwrap();

    // Launch a local network and deploy the contract
    let (provider, wallet) = test_helpers::setup_test_provider_and_wallet().await;

    let id = Contract::deploy(&compiled, &provider, &wallet, TxParameters::default())
        .await
        .unwrap();

    let instance = MyContract::new(id.to_string(), provider, wallet);

    (instance, id)
}

#[tokio::test]
async fn can_get_contract_id() {
    let (instance, _id) = get_contract_instance().await;

    assert_eq!(
        instance
            .small_array_identity([42; 1].to_vec())
            .call()
            .await
            .unwrap()
            .value, [42; 1] 
    );
}
```

I'm testing with `fuels-rs` on master.

Error:
```
thread 'can_get_contract_id' panicked at 'called `Result::unwrap()` on an `Err` value: ContractCallError("RESERV00")', tests/harness.rs:91:14
```

Observations:
* The problem shows up regardless of how many arguments the contract method has or how many elements the array has.
* I don't see the problem when I return a local array (i.e. define an array in the contract method and return it).
* The problem always shows up as long as there's an array in the argument list regardless of whether the array is used or not.
* I think this line needs to handle arrays as well: https://github.com/FuelLabs/fuels-rs/blob/b433c95ad23c4204419b64441c2fed912756777a/packages/fuels-contract/src/contract.rs#L345 (and then specialize for only _large_ arrays once we finalize that optimization in the compiler as well)

**Comment by Salka1988** [comment]
proj test:

```rust
contract;

abi MyContract {
    fn small_array_identity(a: [u64; 1]) -> [u64; 1];
    fn array_hardcoded_in_function_without_parameter() -> [u64; 1];
}

impl MyContract for Contract {
    fn small_array_identity(a: [u64; 1]) -> [u64; 1] {
        a
    }

    fn array_hardcoded_in_function_without_parameter() -> [u64; 1] {
       let a: [u64; 1] = [1];
       a
    }

}
```
When calling abigen function:

```
    abigen!(
        MyContract,
        "packages/fuels-abigen-macro/tests/test_projects/proj/out/debug/proj-abi.json"
    );
```
This contract is generated:

```rust
            pub struct MyContract {
                contract_id: ContractId,
                provider: Provider,
                wallet: LocalWallet,
            }
            impl MyContract {
                pub fn new(contract_id: String, provider: Provider, wallet: LocalWallet) -> Self {
                    let contract_id = ContractId::from_str(&contract_id).unwrap();
                    Self {
                        contract_id,
                        provider,
                        wallet,
                    }
                }
                #[doc = "Calls the contract's `small_array_identity` (0x000000003394437d) function"]
                pub fn small_array_identity(
                    &self,
                    a: ::std::vec::Vec<u64>,
                ) -> ContractCall<::std::vec::Vec<u64>> {
                    Contract::method_hash(
                        &self.provider,
                        self.contract_id,
                        &self.wallet,
                        [0, 0, 0, 0, 51, 148, 67, 125],
                        &[ParamType::Array(Box::new(ParamType::U64), 1)],
                        &[a.into_token()],
                    )
                    .expect("method not found (this should never happen)")
                }
                #[doc = "Calls the contract's `array_hardcoded_in_function_without_parameter` (0x000000008992f8ab) function"]
                pub fn array_hardcoded_in_function_without_parameter(
                    &self,
                ) -> ContractCall<::std::vec::Vec<u64>> {
                    Contract::method_hash(
                        &self.provider,
                        self.contract_id,
                        &self.wallet,
                        [0, 0, 0, 0, 137, 146, 248, 171],
                        &[ParamType::Array(Box::new(ParamType::U64), 1)],
                        &[],
                    )
                    .expect("method not found (this should never happen)")
                }
            }
```

When calling ``array_hardcoded_in_function_without_parameter()`` correct value is returned inside ReturnData type. 
If we add other type of parameter (i tried Tuple ``fn array_hardcoded_in_function_without_parameter(b: (u64, u64))``) to this function, it will also return correct value value (hardcoded one). 

If we call a function that receives an array (in our case "small_array_identity") as a parameter then we have the following situation:

```rust         
->       let receipts = fuel_client.receipts(&tx_id).await?; 
->       FuelClient::query 
->       surf::post(&self.url)
          .run_graphql(q)
          .await
          .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;

Lot of "promises" and then:

->      TxMutation::submit()
->      TxPool::submit_tx()
->      Executor::execute()

After this, we are in fuel-vm:

->      Interpreter::transact()
->      Interpreter::run()
->      Interpreter::run_program()
->      Interpreter::execute()
->      Interpreter::instruction()

```

Inside  ``pub fn instruction(&mut self, instruction: Instruction)`` we trigger error and "Revert(0)" while analysing/computing "hi" and "lo" instructions.

**Comment by digorithm** [comment]
This whole issue is related to this comment here https://github.com/FuelLabs/sway/issues/1631#issuecomment-1133976769. 

TL;DR: no support for arrays as _inputs_ because we're still not properly encoding the fn selector.
